versioned/tiered/rocksdb/src/main/java/org/projectnessie/versioned/rocksdb/RocksBaseValue.java

```diff
@@ -164,6 +175,259 @@ protected static String conditionNotMatchedMessage(Function function) {
+          }
+        } else if (earlierOperator == UpdateFunction.Operator.SET
+            && earlierSubOperator == UpdateFunction.SetFunction.SubOperator.EQUALS
+            && currentOperator == UpdateFunction.Operator.SET
+            && currentSubOperator == UpdateFunction.SetFunction.SubOperator.EQUALS) {
+          if (!currentUpdatePath.equals(earlierUpdatePath) && currentUpdatePath.startsWith(earlierUpdatePath)) {
+            keepUpdateFunction[i] = false;
+          }
+        } else if (earlierOperator == UpdateFunction.Operator.SET
+            && earlierSubOperator == UpdateFunction.SetFunction.SubOperator.EQUALS
+            && currentOperator == UpdateFunction.Operator.SET
+            && currentSubOperator == UpdateFunction.SetFunction.SubOperator.APPEND_TO_LIST) {
+          if (currentUpdatePath.startsWith(earlierUpdatePath)) {
+            keepUpdateFunction[i] = false;
+          }
+        }
+      }
+    }
+
+    final List<UpdateFunction> functionsToApply = new ArrayList<>();
+    for (int i = 0; i < keepUpdateFunction.length; i++) {
+      if (keepUpdateFunction[i]) {
+        functionsToApply.add(updateFunctions.get(i));
+      }
+    }
+    return functionsToApply;
+  }
+
+  // TODO: we might be able to provide implementation here and just have
+  // abstract methods for SET and REMOVE, which will be implemented by the
+  // specific classes L1, L2 etc.
+  /**
+   * Applies an update to the implementing class.
+   * @param updateFunction the update to apply to the implementing class.
+   * @return The PathSegment that should be removed, or null
+   */
+  private ExpressionPath.PathSegment updateWithFunction(UpdateFunction updateFunction) {
+    final ExpressionPath.NameSegment nameSegment = updateFunction.getRootPathAsNameSegment();
+    final String segment = nameSegment.getName();
+    final ExpressionPath.PathSegment childPath = nameSegment.getChild().orElse(null);
+
+    switch (updateFunction.getOperator()) {
+      case REMOVE:
+        if (childPath == null || !fieldIsList(segment, childPath)) {
+          throw new UnsupportedOperationException(
+            String.format("Expression path \"%s\" is not valid", updateFunction.getPath().asString())
+          );
+        }
+
+        return nameSegment;
+      case SET:
+        UpdateFunction.SetFunction setFunction = (UpdateFunction.SetFunction) updateFunction;
+        switch (setFunction.getSubOperator()) {
+          case APPEND_TO_LIST:
+            if (fieldIsList(segment, childPath)) {
+              if (setFunction.getValue().getType() == Entity.EntityType.LIST) {
+                appendToList(segment, childPath, setFunction.getValue().getList());
+              } else {
+                appendToList(segment, childPath, Collections.singletonList(setFunction.getValue()));
+              }
+            } else {
+              throw new UnsupportedOperationException(String.format("The append to list operation is not valid for \"%s\"", segment));
```

Comment:
I'd prefer stylistically to not have the throws in the else, but to have them in the if. This way you don't even need the else, remove nesting, and don't need to refer to above in the code to find out why the exception is triggered.

Reply:
Updated what I could to remove some of the nesting and to move the throws up higher.